server/study_server.js

```diff
@@ -153,56 +153,77 @@ module.exports = {
     },
     submitResult: function (req, res, next) {
         var clean_id = sanitize(req.body.id);
+        var max_retries = 3;
 
-         Study.findOne({_id: clean_id}, function (err, study) {
-            if (err) {
-                res.status(504);
-                logger.error("study_server.js: Error in submit result:", err);
-                res.end(err);
-            } else {
-                //if the study is being previewed, don't record the response
-                if (req.body.resid === "preview"){
-                    res.redirect('/studies');
+        // function to save complete response to study and remove from incomplete responses
+        function saveStudy(study, responseID, retries){
+            var respIdx = study.incompleteResponses.indexOf(responseID);
+            study.completeResponses.push(study.incompleteResponses[respIdx]);
+            study.incompleteResponses.splice(respIdx,1);
+            study.save(function(err) {
+                if (err) {
+                    if (retries < max_retries) {
+                        logger.warn(`Retrying... (${retries + 1}/${max_retries})`);
+                        handleAndSaveResults(retries + 1);
+                    } else {
+                        logger.error("study_server.js: Error saving response to study:", err);
```

Comment:
Can we move this to the start of if condition, worth logging the error every time

Reply:
Sure.